keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1055 @@
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With this proposal, the Kubernetes API server will enable per-request read/write
+timeouts using the `ResponseController` functionality provided by the golang net/http library.
+With the enablement of per-request read/write timeout, the `timeout` filter in the
+kube-apiserver will become redundant, and thus be removed from the server handler chain.
+
+
+## Motivation
+
+The `timeout` filter has been a source of many hard to debug data races in the past. 
+With the removal of the `timeout` filter from the apiserver handler chain, the entire
+request handler will be executed on a single goroutine. Thus it will eliminate 
+the data race conditions that are inherent with the `timeout` filter.
+
+### Goals
+
+The primary goal is to reduce data race conditions by removing the `timeout`
+filter from the apiserver request handler chain.
+
+### Non-Goals
+
+- WATCH requests are outside the scope of this proposal, since currently we
+don't attach any deadline to the `Context` of a WATCH request. An orthogonal
+effort to allocate a deadline to the `Context`of WATCH request(s) is a
+prerequisite, and such effort is outside the scope of this proposal.
+
+- All other long-running requests are also outside the scope of this proposal.
+
+
+## Proposal
+
+A new feature flag `PerHandlerReadWriteTimeout` will be introduced, when enabled:
+- On start-up, the kube-apiserver will remove the `timeout` filter
+  from the request handler chain.
+- To every incoming non long-running request, the apiserver will 
```

Comment:
are our timeout goals better achieved by having a long server-wide timeout that we shorten for short requests (proposed here) or setting a short server-wide timeout that we extend for long running requests?

Reply:
For now, we propose to set a read/write deadline that is equal to the request context deadline, so this is in par with what we have today.

> setting a short server-wide timeout that we extend for long running requests?

long-running requests are outside the scope of this proposal, since we don't set any deadline to the request context of WATCH or other long-running requests.

Reply:
>For now, we propose to set a read/write deadline that is equal to the request context deadline, so this is in par with what we have today.

What are the pros/cons of that choice though.  Defaulting to shorter will make it easier to catch cases of "missed one", but defaulting to longer means we're unlikely to catch cases of "too lenient".

Reply:
sorry, not sure what you are referring to as "missed one".
I guess we can tweak the read timeout, make it shorter than the context deadline, if certain request type has need for it.
making the write deadline shorter than the context deadline will result in the request timing out before the context expires, so it maybe equivalent to making the context deadline shorter.
Not sure if we want to say the request must be done writing before x% of the context deadline.